File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/ToleranceUtils.java
package com.mapbox.services.android.navigation.v5.utils;

import com.mapbox.services.android.navigation.v5.models.StepIntersection;
import com.mapbox.geojson.Point;
import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigationOptions;
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
import com.mapbox.turf.TurfClassification;
import com.mapbox.turf.TurfConstants;
import com.mapbox.turf.TurfMeasurement;

import java.util.ArrayList;
import java.util.List;

public final class ToleranceUtils {

  private ToleranceUtils() {
    // Utils class therefore, shouldn't be initialized.
  }

  /**
   * Reduce the minimumDistanceBeforeRerouting if we are close to an intersection.
   * You can define these values in the navigationOptions
   */
  public static double dynamicRerouteDistanceTolerance(Point snappedPoint,
                                                       RouteProgress routeProgress,
                                                       MapboxNavigationOptions navigationOptions) {
    List<StepIntersection> intersections
      = routeProgress.currentLegProgress().currentStepProgress().intersections();

    if(!intersections.isEmpty()){
      List<Point> intersectionsPoints = new ArrayList<>();
      for (StepIntersection intersection : intersections) {
        intersectionsPoints.add(intersection.location());
      }

      Point closestIntersection = TurfClassification.nearestPoint(snappedPoint, intersectionsPoints);

      if (closestIntersection.equals(snappedPoint)) {
        return navigationOptions.minimumDistanceBeforeRerouting();
      }

      double distanceToNextIntersection = TurfMeasurement.distance(snappedPoint, closestIntersection,
              TurfConstants.UNIT_METERS);

      if (distanceToNextIntersection <= navigationOptions.maneuverZoneRadius()) {
        return navigationOptions.minimumDistanceBeforeRerouting() / 2;
      }
    }

    return navigationOptions.minimumDistanceBeforeRerouting();
  }
}

File: libandroid-navigation/src/commonMain/kotlin/org/maplibre/navigation/core/utils/ToleranceUtils.kt
package org.maplibre.navigation.core.utils

import org.maplibre.geojson.Point
import org.maplibre.navigation.core.navigation.MapLibreNavigationOptions
import org.maplibre.navigation.core.routeprogress.RouteProgress
import org.maplibre.turf.TurfClassification
import org.maplibre.turf.TurfConstants
import org.maplibre.turf.TurfMeasurement
import kotlin.jvm.JvmStatic

object ToleranceUtils {

    /**
     * Reduce the offRouteMinimumDistanceMetersBeforeWrongDirection if we are close to an intersection.
     * You can define these values in the navigationOptions
     */
    @JvmStatic
    fun dynamicOffRouteRadiusTolerance(
        snappedPoint: Point,
        routeProgress: RouteProgress,
        navigationOptions: MapLibreNavigationOptions
    ): Double {
        val intersections = routeProgress.currentLegProgress.currentStepProgress.intersections
        if (!intersections.isNullOrEmpty()) {
            val intersectionsPoints: MutableList<Point> = ArrayList()
            for (intersection in intersections) {
                intersectionsPoints.add(intersection.location)
            }

            val closestIntersection = TurfClassification.nearestPoint(snappedPoint, intersectionsPoints)
            if (closestIntersection == snappedPoint) {
                return navigationOptions.offRouteThresholdRadiusMeters
            }

            val distanceToNextIntersection = TurfMeasurement.distance(
                snappedPoint,
                closestIntersection,
                TurfConstants.UNIT_METERS
            )

            if (distanceToNextIntersection <= navigationOptions.maneuverZoneRadius) {
                return navigationOptions.offRouteThresholdRadiusMeters / 2
            }
        }

        return navigationOptions.offRouteThresholdRadiusMeters
    }
}